lib/galaxy/webapps/galaxy/services/wes.py

```diff
@@ -0,0 +1,985 @@
+
+        Returns:
+            TaskListResponse with paginated tasks
+        """
+        invocation = self._get_invocation(trans, run_id)
+
+        # Get all task logs
+        all_task_logs = self._build_task_logs(trans, invocation)
+
+        # Decode page token to offset
+        offset = _decode_page_token(page_token)
+
+        # Apply pagination
+        end_offset = offset + page_size
+        paginated_logs = all_task_logs[offset:end_offset]
+
+        # Generate next page token if more results exist
+        next_page_token = None
+        if end_offset < len(all_task_logs):
+            next_page_token = _encode_page_token(end_offset)
+
+        return TaskListResponse(
+            task_logs=paginated_logs if paginated_logs else None,
+            next_page_token=next_page_token,
+        )
+
+    def get_run_task(
+        self,
+        trans: ProvidesUserContext,
+        run_id: int,
+        task_id: str,
+    ) -> TaskLog:
+        """Get details for a specific task in a workflow run.
+
+        Args:
+            trans: Galaxy transaction/context
+            run_id: The WES run ID (Galaxy invocation ID)
+            task_id: The task ID (step ID or step.job format)
+
+        Returns:
+            TaskLog object for the specified task
+
+        Raises:
+            exceptions.ObjectNotFound: If task not found
+        """
+        invocation = self._get_invocation(trans, run_id)
+
+        # Get all task logs
+        all_task_logs = self._build_task_logs(trans, invocation)
+
+        # Find matching task by ID
+        for task_log in all_task_logs:
+            if task_log.id == task_id:
+                return task_log
+
+        # Task not found
+        raise exceptions.ObjectNotFound(f"Task {task_id} not found in run {run_id}")
+
+    def _invocation_to_run_id(self, invocation: WorkflowInvocation) -> RunId:
+        """Convert a WorkflowInvocation to a WES RunId."""
+        return self._invocation_id_to_run_id(self._security.encode_id(invocation.id))
+
+    def _invocation_id_to_run_id(self, invocation_id: str) -> RunId:
+        """Convert a WorkflowInvocation ID to a WES RunId."""
+        return RunId(run_id=invocation_id)
+
+    def _get_invocation(
+        self,
+        trans: ProvidesUserContext,
+        run_id: int,
+    ) -> WorkflowInvocation:
+        """Get a workflow invocation by ID.
+
+        Args:
+            trans: Galaxy transaction/context
+            run_id: The invocation ID (unencoded)
+
+        Returns:
+            WorkflowInvocation object
+
+        Raises:
+            exceptions.ObjectNotFound: If invocation not found
+            exceptions.AuthenticationRequired: If user cannot access invocation
+        """
+        invocation = trans.sa_session.query(WorkflowInvocation).filter_by(id=run_id).one_or_none()
+
+        if not invocation:
+            raise exceptions.ObjectNotFound(f"Invocation {run_id} not found")
+
+        if invocation.history.user.id != trans.user.id:
```

Comment:
```suggestion
        if invocation.history.user_id != trans.user.id:
```
and I'm surprised you don't have to `assert invocation.history`. Is `invocation` coming out untyped ?

Reply:
In my IDE invocation has a type but not history.